### PR TITLE
Tech debt 372/packaging polls

### PR DESCRIPTION
**Pull request by Naira21**
This pull request has consisted of such work as:

- build updated UML diargam of Poll module [](https://drive.google.com/drive/folders/1Hpn71PXuXvtRb2Aesmy3lTz-Wdj8bVuQ)
- repacked Poll module inside the project:
       1. organised poll files in one folder — poll;
       2. improved CSS styles of submit and radio buttons, warning message in editing and inputText;
       3. united CreatePoll and EditPoll logics in one — PollForn.

Note: as of today _mounted()_ block in PollForm has **try {} catch {}** logic, and I would like to suggest leaving it in vue component instead of moving to action.ts as it has really easy logic inside — dispatch store data and invoke initDate() method — which is opposed to the code in actions.ts. 
 
Please, review the code at your convenience
Any comments will be much appreciated!

### CODE REVIEW

File: src/polls/models/pollDTO.model.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
-import { PollStatusType, PollsResults } from '@/store/polls/types';
+import { PollStatusType, PollsResults } from '@/polls/store/types';
 import { HouseDTOModel } from '@/houses/models/houseDTO.model';
-import { PollModel, PollModelWithResults } from '@/store/polls/models/poll.model';
+import { PollModel, PollModelWithResults } from './poll.model';
```
Comment (yevhendevbox) [inline src/polls/models/pollDTO.model.ts]: Shouldn't it be imported from:
`'@/polls/models/poll.model'`

instead?

**Review by yevhendevbox — DISMISSED** [review]

File: src/houses/components/ManageHouses.vue [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<template>
+<template>
```
Comment (aleonovdementiev) [inline src/houses/components/ManageHouses.vue]:
```suggestion
<template>
```

File: src/polls/PollForm.vue [hunk 3]
```diff
@@ -12,7 +12,8 @@
           id="poll_title"
           placeholder="Заголовок"
           rows="3"
-          @input="v$.pollData.title.$touch"
+          @blur="v$.pollData.title.$touch"
+          @input="isDisabled = false"
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Will the save button be disabled if we'll type something here, and then delete?

File: src/polls/PollForm.vue [hunk 4]
```diff
@@ -84,22 +124,22 @@
             'p-invalid': v$.pollData.acceptanceCriteria.$error,
           }"
           class="radio-button"
-        ></RadioButton>
-        <label for="item.name">{{ item }}</label>
+        />
+        <label :for="key"> {{ item }} </label>
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Do you need the space after the {{ item }}?

File: src/polls/PollForm.vue [hunk 5]
```diff
@@ -109,26 +149,52 @@
         @click="cancelEditing"
       />
     </div>
+
+    <div v-if="isEditing" class="buttons-container">
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Why do you need two buttons containers? Can we leave only one buttons container but adjust the logic for different cases?

File: src/polls/PollForm.vue [hunk 6]
```diff
@@ -199,9 +292,50 @@ export default defineComponent({
     ChangeDate(value: Date) {
       this.pollData.creationDate = value;
     },
+    onChangeCreationDate() {
+      const dateTomorrow = new Date();
+      dateTomorrow.setDate(dateTomorrow.getDate() + 1);
+      dateTomorrow.setHours(0, 0, 0, 0);
+      if (this.pollData.creationDateInEdition < dateTomorrow) {
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]:
```suggestion

      if (this.pollData.creationDateInEdition < dateTomorrow) {
```

File: src/polls/PollForm.vue [hunk 7]
```diff
@@ -199,9 +292,50 @@ export default defineComponent({
...
+
+        this.finishDate = new Date(this.pollData.creationDateInEdition.getTime() + 14 * 86400000);
+        this.finishDate.setHours(23, 59, 59, 59);
+        this.pollData.completionDate = this.finishDate.toLocaleString('uk-UA').split(',')[0];
+
+        this.isCreationDateHelpActive = false;
+        this.isDisabled = false;
+      }
+    },
+    async initData() {
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]:
```suggestion

    async initData() {
```

File: src/polls/PollForm.vue [hunk 8]
```diff
@@ -199,9 +292,50 @@ export default defineComponent({
...
+      await this.$store.dispatch(
+        `${StoreModuleEnum.pollsStore}/${PollsActionEnum.SET_SELECTED_POLL}`,
+        this.$props.poll.id
+      );
+
+      this.pollData.title = this.selectedPoll?.header ?? '';
+      this.pollData.description = this.selectedPoll?.description ?? 'Повний опис опитування';
+      this.pollData.polledHouses = this.selectedPoll?.polledHouses ?? [];
+      this.pollData.creationDateInEdition = this.selectedPoll?.creationDate.toLocaleString('uk-UA');
+      this.pollData.completionDate = this.selectedPoll?.completionDate.toLocaleString('uk-UA');
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Why do you need this default values set to the poll data if you can set them in data() function?

File: src/polls/PollForm.vue [hunk 9]
```diff
@@ -199,9 +292,50 @@ export default defineComponent({
...
+
+      this.beginDateInEdition = this.selectedPoll?.creationDate || new Date();
+      this.finishDate = this.selectedPoll?.completionDate;
+    },
+    submitPollForm() {
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]:
```suggestion

    submitPollForm() {
```

File: src/polls/PollForm.vue [hunk 10]
```diff
@@ -199,9 +292,50 @@ export default defineComponent({
...
+      if (this.isEditing) {
+        this.editPoll();
+      } else {
+        this.createPoll();
+      }
+    },
     async createPoll() {
       const isFormValid = await this.v$.$validate();
       if (!isFormValid) {
+        console.log('i am in isFormValid');
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Do you need this console.log?

File: src/polls/PollForm.vue [hunk 11]
```diff
@@ -224,18 +358,42 @@ export default defineComponent({
...
       this.$emit('create-poll');
     },
+
+    async editPoll() {
+      const poll = {
+        header: this.pollData.title,
+        description: this.pollData.description,
+        creationDate: new Date(this.beginDateInEdition.toLocaleString('en-US')).toISOString(),
+        completionDate: new Date(this.finishDate.toLocaleString('en-US')).toISOString(),
+        status: this.$props.poll.status,
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Maybe this.poll will work instead of this.$props.poll

File: src/polls/PollForm.vue [hunk 12]
```diff
@@ -244,26 +402,28 @@ export default defineComponent({
     },
   },
   mounted() {
-    this.$store.dispatch(`${StoreModuleEnum.housesStore}/${HousesActionsEnum.SET_HOUSES}`, this.$props.cooperationId);
+    if (this.$props.isEditing) {
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Same here

File: src/polls/PollForm.vue [hunk 13]
```diff
@@ -244,26 +402,28 @@ export default defineComponent({
...
+      try {
+        this.$store.dispatch(`${StoreModuleEnum.housesStore}/${HousesActionsEnum.SET_HOUSES}`, this.cooperationIdEdit);
+        this.initData();
+      } catch {
+        console.log('error was caught during mounting BaseCooperationPoll');
+      }
+    } else if (!this.$props.isEditing) {
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: Do you need this condition in 'else if' or just 'else' will do?

File: src/polls/PollForm.vue [hunk 14]
```diff
@@ -244,26 +402,28 @@ export default defineComponent({
...
+      this.$store.dispatch(`${StoreModuleEnum.housesStore}/${HousesActionsEnum.SET_HOUSES}`, this.$props.cooperationId);
```
Comment (aleonovdementiev) [inline src/polls/PollForm.vue]: this.cooperationId?

File: src/polls/models/poll.model.ts [hunk 15]
```diff
@@ -1,6 +1,6 @@
-import { PollStatusType, PollsResults } from '@/store/polls/types';
+import { PollStatusType, PollsResults } from '@/polls/store/types';
```
Comment (aleonovdementyev) [inline src/polls/models/poll.model.ts:1]: It's better to move 'models' folder to the 'store' folder

File: src/polls/models/poll.model.ts [hunk 16]
```diff
@@ -1,6 +1,6 @@
...
 import { HouseModel } from '@/houses/models/house.model';
-import { PollDTOModel, PollDTOModelWithResults } from '@/store/polls/models/pollDTO.model';
+import { PollDTOModel, PollDTOModelWithResults } from '@/polls/models/pollDTO.model';
```
Comment (aleonovdementyev) [inline src/polls/models/poll.model.ts:3]: It's better to move 'models' folder to the 'store' folder

File: src/polls/models/pollDTO.model.ts (hunk 1 above)
Comment (aleonovdementyev) [reply src/polls/models/pollDTO.model.ts:3]: It's better to put it to the '@/poll/store/model/poll.model'

**Comment by aleonovdementiev** [comment]
It's better to name the root folder for this package - 'poll', not 'polls', because we're discribing a single realization for poll, but not many instances of polls

File: src/polls/models/pollDTO.model.ts (hunk 1 above)
Comment (Naira21) [reply src/polls/models/pollDTO.model.ts]: It looks better (unficied) with '@/polls/models/poll.model'. Thank you, corrected!

File: src/polls/PollForm.vue (hunk 3 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: The button becomes disabled after typing and deleting any text here if you are in editing. If you are in creation. it remains active all the time whatever fields you fill/don't fill

Comment (Naira21) [reply src/polls/PollForm.vue]: @aleonovdementyev  Should it be disabled in poll creation as well as in poll edition?

File: src/polls/PollForm.vue (hunk 4 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: definitely not )
thank you for noticing!

File: src/polls/PollForm.vue (hunk 10 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: Absolutely not (missed this console.log before making PR)

File: src/polls/PollForm.vue (hunk 13 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: I agree with you, "else" is enough here and it is working pretty well
I just wanted to single out and clearly indicate the type so that there was no unpredictable option. But as we are using this logic only in polls, therefore there is no other situation and no need in "else if" construction

File: src/polls/PollForm.vue (hunk 14 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: yes, corrected

File: src/polls/PollForm.vue (hunk 8 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: yes, you're right, totally agree with your suggestion

File: src/polls/PollForm.vue (hunk 3 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: updated: we have found common ground to leave it in way it works now

File: src/polls/PollForm.vue (hunk 8 above)
Comment (Naira21) [reply src/polls/PollForm.vue]: @aleonovdementyev, I was wondering to know if I may not put 'Повний опис опитування' as dafault in data()?
There are some issues:
1) technical: ternary operator like this !this.isEditing ? '' : 'Повний опис опитування' (and some other its versions) does not work in option of description in data(). I also attempted to pass via computed expression (for example), but it was failed.
2) logical: according to this default meaning of description, if there is no desc in polledData, please show 'Повний опис опитування'. But it is impossible to be empty because the poll form can't be sumbitted in this case, it becomes invalid.
Also, it is related to all other default values put in lines 319-321 — any data must be written there.
Here is another problem: we do not receive data from polledData, therefore I just suppose that previous engineer put default values to avoid errors.
To sum up, I would suggest not mentioning 'Повний опис опитування' as default in data() as we are passing in all other parameters initial (=empty) values.


Also, @yevhendevbox and @anatoliy-stepanchenko, if you have any other ideas how to put conditional expression in value in data(), I will be so thankful for your suggestions!

File: src/poll/PollForm.vue [hunk 17]
```diff
@@ -196,9 +278,55 @@ export default defineComponent({
...
+      dateTomorrow.setDate(dateTomorrow.getDate() + 1);
+      dateTomorrow.setHours(0, 0, 0, 0);
+
+      if (this.pollData.creationDateInEdition < dateTomorrow) {
+        this.isDisabled = true;
+        this.isCreationDateHelpActive = true;
+      } else {
+        this.beginDateInEdition.setHours(0, 0, 0, 0);
+
+        this.finishDate = new Date(this.pollData.creationDateInEdition.getTime() + 14 * 86400000);
```
Comment (aleonovdementiev) [inline src/poll/PollForm.vue]: It's better to extract 14*86400000 into constant to know what this numbers are.

**Review by aleonovdementiev — DISMISSED** [review]

**Review by yevhendevbox — DISMISSED** [review]

**Review by anatoliy-stepanchenko — DISMISSED** [review]

**Review by yevhendevbox — DISMISSED** [review]

**Review by anatoliy-stepanchenko — APPROVED** [review]

**Review by yevhendevbox — APPROVED** [review]
